### DISCUSSION TITLE
Add Battery Monitor for Hyperx Cloud Flight S (macOS Sonoma 14.4.1 - Chip M2)

### DISCUSSION THREAD

**Discussion by AndersonVD** [Sapd/HeadsetControl#342] [General]
Hello fellas, I came here from this issue https://github.com/srn/hyperx-cloud-flight-wireless/issues/4, my goal is tested and make some changes to work just like https://hyperx.com/pages/ngenuity, adding this software function like mute, play pause, surround 7.1 on off, and other. I want to use HeadsetControl API for that. 

I have done a lot of research and this how I got so far:
Someone who did it before using Node
(https://github.com/srn/hyperx-cloud-flight-wireless/)

Software I'm using for debug and get all info from USB Port and I already have a lot of data that can be used, I just don't know how at the moment. 
(https://ww5.touch-base.com/documentation/USBMacProber.htm)

When a run `./headseatcontrol -o json` I got this error message "Error retrieving battery status. Error: IOHIDDeviceSetReport failed: (0xE0005000) unknown error code"
<details>

<summary>full log</summary>

``` json
{
  "name": "HeadsetControl",
  "version": "3.0.0-4-g9af6929-modified",
  "api_version": "1.0",
  "hidapi_version": "0.14.0",
  "device_count": 1,
  "devices": [
    {
      "status": "partial",
      "device": "HyperX Cloud Flight Wireless",
      "vendor": "Kingston",
      "product": "HyperX Cloud Flight S",
      "id_vendor": "0x0951",
      "id_product": "0x16ea",
      "capabilities": [
        "CAP_BATTERY_STATUS"
      ],
      "capabilities_str": [
        "battery"
      ],
      "errors": {
        "battery": "Error retrieving battery status. Error: IOHIDDeviceSetReport failed: (0xE0005000) unknown error code"
      }
    }
  ]
}

```
</details>

In this link have something about the problem, are buffer sized be different on Mac
(https://stackoverflow.com/questions/42226893/cant-send-hid-feature-reports-of-more-than-64-bytes-on-mac)

I want a help to discovery how I can implement this new features, I don't have exp programming in C, but I will do my best to make this plugin come alive, I'm a python, dart programmer. If anyone could help, will be enjoyable!

**Comment by Sapd** [comment]
Maybe you could try truncating the 0 of the hid_write. Like this, but not sure if this would make a difference:

```
headsetcontrol --dev -- --device 0x0951:0x16ea --send "0x21, 0xff, 0x05" --receive
```

**Reply by AndersonVD** [reply]
Thanks for the quick answer, I don't know if I should put this in the hyperx_cloud_flight.c file, as just running it in the terminal gives me the same error message. 
I'll let you know if I find any kind of solution

**Reply by Sapd** [reply]
No its a command for the terminal where you can basically do (with the --dev parameter) everything via command line instead of programming it. Because you mentioned buffer size I thought cutting the ending 0's might help (which I did with the command).

Practically I think you need to look why or what is different. Basically, capture with WireShark on Windows the same command, then see if something is different.
